File: app/src/main/java/fr/xgouchet/packageexplorer/ui/fragments/AppListFragment.java
package fr.xgouchet.packageexplorer.ui.fragments;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import butterknife.BindView;
import butterknife.ButterKnife;
import fr.xgouchet.packageexplorer.R;
import fr.xgouchet.packageexplorer.model.Apps;
import fr.xgouchet.packageexplorer.ui.adapters.AppsAdapter;
import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * @author Xavier Gouchet
 */
public class AppListFragment extends Fragment implements SearchView.OnQueryTextListener {

    @BindView(R.id.app_list)
    RecyclerView recyclerView;

    private final AppsAdapter appsAdapter = new AppsAdapter();

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);

        appsAdapter.clear();

        Observable.create(new Apps.List(getActivity()))
                .subscribeOn(Schedulers.io())
                .buffer(32)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(appsAdapter);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_list, container, false);
        ButterKnife.bind(this, root);

        recyclerView.setAdapter(appsAdapter);

        return root;
    }

    @Override
    public void onResume() {
        super.onResume();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);

        inflater.inflate(R.menu.search_menu, menu);

        final MenuItem item = menu.findItem(R.id.action_search);
        final SearchView searchView = (SearchView) MenuItemCompat.getActionView(item);
        searchView.setOnQueryTextListener(this);
    }


    @Override
    public boolean onQueryTextSubmit(String query) {

        return false;
    }

    @Override
    public boolean onQueryTextChange(String newText) {
        if (newText == null) {
            appsAdapter.updateFilter("");
        } else {
            appsAdapter.updateFilter(newText);
        }
        return true;
    }

}

File: app/src/main/java/fr/xgouchet/packageexplorer/applist/AppListFragment.kt
package fr.xgouchet.packageexplorer.applist

import android.app.Activity
import android.content.Intent
import android.content.pm.ResolveInfo
import android.net.Uri
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.widget.SearchView
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import com.google.android.gms.oss.licenses.OssLicensesMenuActivity
import fr.xgouchet.packageexplorer.R
import fr.xgouchet.packageexplorer.about.AboutActivity
import fr.xgouchet.packageexplorer.applist.sort.AppSort
import fr.xgouchet.packageexplorer.details.apk.ApkDetailsActivity
import fr.xgouchet.packageexplorer.launcher.LauncherDialog
import fr.xgouchet.packageexplorer.ui.adapter.BaseAdapter
import fr.xgouchet.packageexplorer.ui.mvp.list.BaseListFragment
import io.reactivex.functions.Consumer


class AppListFragment
    : BaseListFragment<AppViewModel, AppListPresenter>(),
        Consumer<AppViewModel> {

    override val adapter: BaseAdapter<AppViewModel> = AppAdapter(this, this)
    override val isFabVisible: Boolean = false
    override val fabIconOverride: Int? = null

    // region Fragment

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setHasOptionsMenu(true)
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater?.inflate(R.menu.app_list, menu)

        menu?.findItem(R.id.action_search)?.let {
            val searchView = it.actionView as SearchView
            searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
                override fun onQueryTextSubmit(query: String?): Boolean = false

                override fun onQueryTextChange(newText: String?): Boolean {
                    presenter.setFilter(newText ?: "")
                    return true
                }
            })
        }

        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onPrepareOptionsMenu(menu: Menu?) {
        super.onPrepareOptionsMenu(menu)
        val systemAppsVisible = presenter.areSystemAppsVisible()
        menu?.apply {
            findItem(R.id.hide_system_apps).isVisible = systemAppsVisible
            findItem(R.id.show_system_apps).isVisible = !systemAppsVisible
        }
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {

        when (item?.itemId) {
            R.id.sort_by_title -> presenter.setSort(AppSort.TITLE)
            R.id.sort_by_package_name -> presenter.setSort(AppSort.PACKAGE_NAME)
            R.id.sort_by_install_time -> presenter.setSort(AppSort.INSTALL_TIME)
            R.id.sort_by_update_time -> presenter.setSort(AppSort.UPDATE_TIME)
            R.id.open_apk -> {
                val intent = Intent(Intent.ACTION_OPEN_DOCUMENT)
                intent.type = APK_MIME_TYPE
                intent.addCategory(Intent.CATEGORY_OPENABLE)
                startActivityForResult(intent, OPEN_APK_REQUEST)
            }
            R.id.hide_system_apps -> presenter.setSystemAppsVisible(false)
            R.id.show_system_apps -> presenter.setSystemAppsVisible(true)
            R.id.about -> {
                startActivity(Intent(activity, AboutActivity::class.java))
            }
            R.id.licenses -> {
                OssLicensesMenuActivity.setActivityTitle(getString(R.string.activity_title_licenses))
                startActivity(Intent(activity, OssLicensesMenuActivity::class.java))
            }
            else -> return super.onOptionsItemSelected(item)
        }
        return true
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (requestCode) {
            OPEN_APK_REQUEST -> onOpenApkResult(resultCode, data)
        }
    }

    // endregion

    // region Displayer

    fun promptActivity(resolvedInfos: List<ResolveInfo>) {
        val supportFragmentManager = activity?.supportFragmentManager ?: return
        val transaction = supportFragmentManager.beginTransaction()
        LauncherDialog.withData(resolvedInfos)
                .show(transaction, null)
    }

    // endregion

    // region Consumer

    override fun accept(t: AppViewModel) {
        presenter.openApplication(t.packageName)
    }

    // endregion

    // region Private

    private fun onOpenApkResult(resultCode: Int, resultData: Intent?) {
        if (resultCode == Activity.RESULT_OK) {
            val uri: Uri? = resultData?.data
            if (uri != null) {
                val intent = Intent(context, ApkDetailsActivity::class.java)
                intent.data = uri
                startActivity(intent)
            }
        }
    }

    // endregion

    companion object {
        const val APK_MIME_TYPE = "application/vnd.android.package-archive"
        const val OPEN_APK_REQUEST = 1
    }
}
